lib/lookup/src/lookup_v2/jit.rs

```diff
@@ -0,0 +1,216 @@
+use crate::lookup_v2::{BorrowedSegment, Path};
+use std::str::CharIndices;
+
+#[derive(Clone)]
+pub struct JitPath<'a> {
+    path: &'a str,
+}
+
+impl JitPath<'_> {
+    pub fn new(path: &str) -> JitPath {
+        JitPath { path }
+    }
+}
+
+/// This is essentially an iterator over a `JitPath`.
+pub struct JitLookup<'a> {
+    path: &'a str,
+    chars: CharIndices<'a>,
+    state: JitState,
+}
+
+impl<'a> JitLookup<'a> {
+    pub fn new(path: &'a str) -> Self {
+        Self {
+            chars: path.char_indices(),
+            path,
+            state: JitState::Start,
+        }
+    }
+}
+
+impl<'a> Path<'a> for JitPath<'a> {
+    type Iter = JitLookup<'a>;
+
+    fn segment_iter(&self) -> Self::Iter {
+        JitLookup::new(self.path)
+    }
+}
+
+enum JitState {
+    Start,
+    Continue,
+    Dot,
+    IndexStart,
+    Index { value: usize },
+    Field { start: usize },
+    Quote { start: usize },
+    End,
+}
+
+impl<'a> Iterator for JitLookup<'a> {
+    type Item = BorrowedSegment<'a>;
+
+    fn next(&mut self) -> Option<BorrowedSegment<'a>> {
+        loop {
+            match self.chars.next() {
+                None => {
+                    let result = match self.state {
+                        JitState::Start => Some(BorrowedSegment::Invalid),
+                        JitState::Continue => None,
+                        JitState::Dot => None,
+                        JitState::IndexStart => Some(BorrowedSegment::Invalid),
+                        JitState::Index { .. } => Some(BorrowedSegment::Invalid),
+                        JitState::Field { start } => {
+                            Some(BorrowedSegment::Field(&self.path[start..]))
+                        }
+                        JitState::Quote { .. } => Some(BorrowedSegment::Invalid),
+                        JitState::End => None,
+                    };
+                    self.state = JitState::End;
+                    return result;
+                }
+                Some((index, c)) => {
+                    let (result, state) = match self.state {
+                        JitState::Start | JitState::Continue => match c {
+                            '.' => (None, JitState::Dot),
+                            'A'..='Z' | 'a'..='z' | '_' | '0'..='9' => {
```

Comment:
nit: since these patterns are re-used, and might be expanded in the future, it might make sense to extract them into a const.

Reply:
Also, note that VRL supports unescaped `@` in paths because it is such a common character in fields. So, this is valid: `.@timestamp`.

Reply:
Interesting. The `lookup` crate didn't support this, so I missed it. Easy to add though

Reply:
Yeah, the timeline of this was something like: VRL introduced the new path parser, lookup crate took the path parsing parts of the VRL parser, and moved it into a new crate, then VRL extended path to allow `@`, but we never migrated that over to lookup.

I don't _think_ there's anything else we changed to the path parsing logic in-between, so adding this (and supporting escaping quotes, as discussed elsewhere) is all we need to do.

Reply:
done